### PR TITLE
Implement function `json_metadata` in `decl_module!`

### PR DESCRIPTION
**Pull request by bkchr**
The new function `json_metadata` returns the module declaration
as a json string.

Progress on: #535

### CODE REVIEW

**Comment by gavofyork** [comment]
good stuff!

File: substrate/runtime-support/src/dispatch.rs [hunk 1]
```diff
@@ -558,3 +568,241 @@ macro_rules! impl_outer_dispatch_common {
...
+/// Implement the `json_metadata` function.
+#[macro_export]
+macro_rules! __impl_json_metadata {
+	(
+		impl for $mod_type:ident<$trait_instance:ident: $trait_name:ident>;
+		$($rest:tt)*
+	) => {
+		impl<$trait_instance: $trait_name> $mod_type<$trait_instance> {
+			pub fn json_metadata() -> &'static str {
+			    concat!(r#"{ "name": ""#, stringify!($mod_type), r#"", "calls": ["#,
```
Comment (pepyakin) [inline substrate/runtime-support/src/dispatch.rs]: spaces

File: substrate/runtime-support/src/dispatch.rs [hunk 2]
```diff
@@ -558,3 +568,241 @@ macro_rules! impl_outer_dispatch_common {
...
+						, $param_name:ident : $param:ty
+					)*
+				) -> $result:ty
+				= $id:expr ;
+			)*
+		}
+		$($rest:tt)*
+	) => {
+		concat!($prefix_str, " ",
+		    r#"{ "name": ""#, stringify!($call_type),
```
Comment (pepyakin) [inline substrate/runtime-support/src/dispatch.rs]: spaces

File: substrate/runtime-support/src/dispatch.rs [hunk 3]
```diff
@@ -558,3 +568,241 @@ macro_rules! impl_outer_dispatch_common {
...
+						$param_name:ident : $param:ty
+					),*
+				) -> $result:ty
+				= $id:expr ;
+			)*
+		}
+		$($rest:tt)*
+	) => {
+		concat!($prefix_str, " ",
+		    r#"{ "name": ""#, stringify!($call_type),
```
Comment (pepyakin) [inline substrate/runtime-support/src/dispatch.rs]: spaces

File: substrate/runtime-support/src/dispatch.rs [hunk 4]
```diff
@@ -558,3 +568,241 @@ macro_rules! impl_outer_dispatch_common {
...
+				    __calls_to_json!(""; $($rest)*), " ] }")
+			}
+		}
+	}
+}
+
+/// Convert the list of calls into their JSON representation, joined by ",".
+#[macro_export]
+macro_rules! __calls_to_json {
+    // WITHOUT AUX
```
Comment (pepyakin) [inline substrate/runtime-support/src/dispatch.rs]: spaces

File: substrate/runtime-support/src/dispatch.rs [hunk 5]
```diff
@@ -558,3 +568,241 @@ macro_rules! impl_outer_dispatch_common {
...
+#[macro_export]
+macro_rules! __impl_json_metadata {
+	(
+		impl for $mod_type:ident<$trait_instance:ident: $trait_name:ident>;
+		$($rest:tt)*
+	) => {
+		impl<$trait_instance: $trait_name> $mod_type<$trait_instance> {
+			pub fn json_metadata() -> &'static str {
+			    concat!(r#"{ "name": ""#, stringify!($mod_type), r#"", "calls": ["#,
+				    __calls_to_json!(""; $($rest)*), " ] }")
```
Comment (pepyakin) [inline substrate/runtime-support/src/dispatch.rs]: spaces

**Pull request merged**